### PR TITLE
Enhance audio transcription with temp files and add GPU support

### PR DESCRIPTION
**Pull request by ItsSimko**
Issues: #15 #21
## Summary by Sourcery

Update the audio transcription endpoint to handle audio files using temporary files and introduce GPU support in the Docker Compose configuration for improved performance.

Enhancements:
- Improve audio file handling by using temporary files for transcription instead of in-memory buffers.

Build:
- Add GPU support in the Docker Compose configuration by reserving NVIDIA devices for the speech-container service.

## Summary by Sourcery

Enhance the audio transcription endpoint by using temporary files for audio processing and introduce GPU support in the Docker Compose configuration to improve performance.

Enhancements:
- Improve audio file handling by using temporary files for transcription instead of in-memory buffers.

Build:
- Add GPU support in the Docker Compose configuration by reserving NVIDIA devices for the speech-container service.

### CODE REVIEW

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: server.py [hunk 1]
```diff
@@ -172,25 +172,23 @@ async def transcribe_audio(
...
     except Exception as e:
-        logging.error(f"Error processing audio file: {e}")
-        raise HTTPException(
-            status_code=500, detail=f"Error processing audio file: {e}"
-        ) from e
-
-    return response_data
+        raise HTTPException(status_code=500, detail=str(e))
+    finally:
+        if 'temp_path' in locals():
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:190]: Elaborate further?

File: server.py [hunk 2]
```diff
@@ -172,25 +172,23 @@ async def transcribe_audio(
...
-        audio_data, sample_rate = librosa.load(
-            audio_buffer, sr=None
-        )  # sr=None to keep the original sample rate
-
-        # Process the file with Whisper using the in-memory buffer
-        result = MODEL.transcribe(
-            audio_data
-        )  # Assuming 'model' can handle file-like objects
-        response_data = {"text": result["text"]}
+        # Save to temporary file
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:175]: For some reason, whisper does not like in mem audio buffer. It caused it not to transcribe properly and caused significantly longer transcribing times

File: server.py [hunk 3]
```diff
@@ -220,12 +218,17 @@ async def startup_event():
 
     # Load the Whisper model using the specified model name
     global MODEL
-    MODEL = whisper.load_model(args["whispermodel"])
+    device = "cpu"
+    
+    if args["use_gpu"] == "True":
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py]: https://github.com/ClinicianFOCUS/speech2text-container/pull/20/commits/fa8f7e21cf6e8e802821c036212fd10dc3c3225c

File: server.py [hunk 4]
```diff
@@ -220,12 +218,17 @@ async def startup_event():
...
+        device = "cuda"
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py]: https://github.com/ClinicianFOCUS/speech2text-container/pull/20/commits/b77dcb642762add061b5a529bc249e1df4520dde

File: server.py [hunk 5]
```diff
@@ -172,25 +172,23 @@ async def transcribe_audio(
...
+            temp_path = temp_file.name
+            # Transcribe using temporary file
+            result = MODEL.transcribe(
+                temp_path,
+            )
+            temp_file.close() # close the file so it is deleted
+        
+        return {"text": result["text"]}
+    
     except Exception as e:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:186]: Done

File: server.py [hunk 6]
```diff
@@ -172,25 +172,23 @@ async def transcribe_audio(
         )
 
     try:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:174]: https://github.com/ClinicianFOCUS/speech2text-container/pull/20/commits/95b90099c1778ed48a124d4dc0fb68531e2396a2

File: server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:190]: @sourcery-ai elaborate?

File: server.py [hunk 7]
```diff
@@ -220,12 +218,14 @@ async def startup_event():
...
     # Load the Whisper model using the specified model name
     global MODEL
-    MODEL = whisper.load_model(args["whispermodel"])
+    device = "cuda" if args["use_gpu"] is True else "cpu"
 
     # Check and retrieve the API key\
     global SESSION_API_KEY
     SESSION_API_KEY = check_api_key()
 
+    MODEL = whisper.load_model(name=args["whispermodel"], device=device)
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py:227]: I think the flag is better and gives user more control. Im open to to suggestions. Do we want to force start it on CPU if gpu unavail?

File: server.py [hunk 8]
```diff
@@ -172,25 +172,23 @@ async def transcribe_audio(
...
-            audio_buffer, sr=None
-        )  # sr=None to keep the original sample rate
-
-        # Process the file with Whisper using the in-memory buffer
-        result = MODEL.transcribe(
-            audio_data
-        )  # Assuming 'model' can handle file-like objects
-        response_data = {"text": result["text"]}
+        # Save to temporary file
+        with tempfile.NamedTemporaryFile(delete=False, suffix='.wav') as temp_file:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply server.py]: https://github.com/ClinicianFOCUS/speech2text-container/pull/20/commits/30e59c5ecb3cf0dfaada0fdebf6d909c73e27527

File: server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply server.py:227]: Should still be a handled error I think.

**Review by yingbull — APPROVED** [review]

**Pull request merged**